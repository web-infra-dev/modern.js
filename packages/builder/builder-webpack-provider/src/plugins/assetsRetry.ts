import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
