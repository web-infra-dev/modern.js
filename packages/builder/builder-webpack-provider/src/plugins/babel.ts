import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
