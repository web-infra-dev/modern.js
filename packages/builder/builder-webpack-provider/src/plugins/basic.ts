import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
