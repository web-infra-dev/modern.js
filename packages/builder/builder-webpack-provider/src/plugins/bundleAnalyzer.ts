import { BuilderPlugin } from '../types';

export function builderPluginBundleAnalyzer(): BuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyWebpackChain(
        async (chain, { CHAIN_ID, getCompiledPath, target }) => {
          const config = api.getNormalizedConfig();
          // There are two ways to open the bundle analyzer:
          // 1. Set environment variable `BUNDLE_ANALYZE`
          // 2. Set performance.bundleAnalyze config
          if (
            !process.env.BUNDLE_ANALYZE &&
            !config.performance.bundleAnalyze
          ) {
            return;
          }
          const { default: BundleAnalyzer } = await import(
            getCompiledPath('webpack-bundle-analyzer')
          );
          chain
            .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
            .use(BundleAnalyzer.BundleAnalyzerPlugin, [
              {
                analyzerMode: 'static',
                openAnalyzer: false,
                reportFilename: `report-${target}.html`,
                ...(config.performance.bundleAnalyze || {}),
              },
            ]);
        },
      );
    },
  };
}
