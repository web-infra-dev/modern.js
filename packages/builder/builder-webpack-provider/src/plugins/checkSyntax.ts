import {
  BuilderContext,
  BuilderTarget,
  DEFAULT_BROWSERSLIST,
  getBrowserslistWithDefault,
} from '@modern-js/builder-shared';
import type { BuilderPlugin, NormalizedConfig } from '../types';

export function builderPluginCheckSyntax(): BuilderPlugin {
  return {
    name: 'builder-plugin-check-syntax',

    setup(api) {
      api.modifyWebpackChain(async (chain, { isProd, target }) => {
        const config = api.getNormalizedConfig();
        const { checkSyntax } = config.security;

        if (
          !isProd ||
          ['node', 'web-worker'].includes(target) ||
          !checkSyntax
        ) {
          return;
        }

        const targets = await getCheckTargets(
          api.context,
          config,
          target,
          checkSyntax,
        );
        const { CheckSyntaxPlugin } = await import(
          '../webpackPlugins/CheckSyntaxPlugin'
        );
        chain.plugin(CheckSyntaxPlugin.name).use(CheckSyntaxPlugin, [
          {
            targets,
            exclude:
              typeof checkSyntax === 'object' ? checkSyntax.exclude : undefined,
          },
        ]);
      });
    },
  };
}

async function getCheckTargets(
  builderContext: BuilderContext,
  builderConfig: NormalizedConfig,
  builderTarget: BuilderTarget,
  checkSyntax: { targets: string[] } | true,
) {
  if (checkSyntax === true) {
    const browserslist = await getBrowserslistWithDefault(
      builderContext.rootPath,
      builderConfig,
      builderTarget,
    );

    return browserslist || DEFAULT_BROWSERSLIST[builderTarget];
  }
  return checkSyntax.targets;
}
