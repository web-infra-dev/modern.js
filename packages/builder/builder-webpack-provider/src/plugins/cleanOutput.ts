import type { BuilderPlugin } from '../types';

export const PluginCleanOutput = (): BuilderPlugin => ({
  name: 'builder-plugin-clean-output',

  setup(api) {
    const clean = async () => {
      const config = api.getNormalizedConfig();

      if (config.output.cleanDistPath) {
        const { emptyDir } = await import('@modern-js/utils');
        const { distPath } = api.context;
        await emptyDir(distPath);
      }
    };

    api.onBeforeBuild(clean);
    api.onBeforeStartDevServer(clean);
  },
});
