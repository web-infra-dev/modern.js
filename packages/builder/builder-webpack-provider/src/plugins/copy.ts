import type { CopyPluginOptions, BuilderPlugin } from '../types';

export const builderPluginCopy = (): BuilderPlugin => ({
  name: 'builder-plugin-copy',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const { copy } = config.output;

      if (!copy) {
        return;
      }

      const { default: CopyPlugin } = await import(
        '../../compiled/copy-webpack-plugin'
      );

      const options: CopyPluginOptions = Array.isArray(copy)
        ? { patterns: copy }
        : copy;

      chain.plugin(CHAIN_ID.PLUGIN.COPY).use(CopyPlugin, [options]);
    });

    api.modifyWebpackConfig(async config => {
      const copyPlugin = config.plugins?.find(
        item => item.constructor.name === 'CopyPlugin',
      ) as unknown as CopyPluginOptions;

      if (copyPlugin) {
        const { fs } = await import('@modern-js/utils');

        // If the pattern.context directory not exists, we should remove CopyPlugin.
        // Otherwise the CopyPlugin will cause the webpack to re-compile.
        const isContextNotExists = copyPlugin.patterns.every(
          pattern =>
            typeof pattern !== 'string' &&
            pattern.context &&
            !fs.existsSync(pattern.context),
        );
        if (isContextNotExists) {
          config.plugins = config.plugins?.filter(
            item => item.constructor.name !== 'CopyPlugin',
          );
        }
      }
    });
  },
});
