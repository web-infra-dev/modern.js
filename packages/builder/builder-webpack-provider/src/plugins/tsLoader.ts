import { TS_REGEX } from '@modern-js/builder-shared';
import { BuilderPlugin } from '../types';
import { applyScriptCondition, getUseBuiltIns } from './babel';

// Declare `ScriptTarget` and `ModuleKind` manualy to avoid high cost of typescript import
enum ScriptTarget {
  ESNext = 99,
}

enum ModuleKind {
  ESNext = 99,
}

export const PluginTsLoader = (): BuilderPlugin => {
  return {
    name: 'builder-plugin-ts-loader',
    setup(api) {
      const config = api.getBuilderConfig();
      if (!config.tools?.tsLoader) {
        return;
      }

      const { framework, rootPath } = api.context;
      const babelLoaderOptions = {
        presets: [
          [
            require.resolve('@modern-js/babel-preset-app'),
            {
              metaName: framework,
              appDirectory: rootPath,
              target: 'client',
              useTsLoader: true,
              useBuiltIns: getUseBuiltIns(config),
              userBabelConfig: config.tools.babel,
            },
          ],
        ],
      };

      const includes: Array<string | RegExp> = [];
      const excludes: Array<string | RegExp> = [];

      const tsLoaderUtils = {
        addIncludes(items: string | RegExp | (string | RegExp)[]) {
          if (Array.isArray(items)) {
            includes.push(...items);
          } else {
            includes.push(items);
          }
        },
        addExcludes(items: string | RegExp | (string | RegExp)[]) {
          if (Array.isArray(items)) {
            excludes.push(...items);
          } else {
            excludes.push(items);
          }
        },
      };
      api.modifyWebpackChain(async (chain, { getCompiledPath }) => {
        const { CHAIN_ID, applyOptionsChain } = await import(
          '@modern-js/utils'
        );
        const tsLoaderOptions = applyOptionsChain(
          {
            compilerOptions: {
              target: ScriptTarget.ESNext,
              module: ModuleKind.ESNext,
            },
            transpileOnly: true,
            allowTsInNodeModules: true,
          },
          config.tools?.tsLoader || {},
          tsLoaderUtils,
        );
        const rule = chain.module.rule(CHAIN_ID.RULE.TS);

        applyScriptCondition(rule, config, api.context, includes, excludes);

        rule
          .test(TS_REGEX)
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelLoaderOptions)
          .end()
          .use(CHAIN_ID.USE.TS)
          .loader(require.resolve('ts-loader'))
          .options(tsLoaderOptions);
      });
    },
  };
};
