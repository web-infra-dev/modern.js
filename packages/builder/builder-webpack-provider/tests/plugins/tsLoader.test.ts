import { expect, describe, it } from 'vitest';
import { builderPluginTsLoader } from '@/plugins/tsLoader';
import { createStubBuilder } from '@/stub';

describe('plugins/tsLoader', () => {
  it('should set ts-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginTsLoader()],
      builderConfig: {
        tools: {
          tsLoader: {},
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should set include/exclude', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginTsLoader()],
      builderConfig: {
        tools: {
          tsLoader(options, { addIncludes, addExcludes }) {
            addIncludes(['src/**/*.ts']);
            addExcludes(['src/**/*.js']);
            return options;
          },
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
