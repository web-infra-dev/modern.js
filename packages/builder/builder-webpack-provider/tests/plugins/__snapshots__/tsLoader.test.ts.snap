// Vitest Snapshot v1

exports[`plugins/tsLoader > should set include/exclude 1`] = `
{
  "module": {
    "rules": [
      {
        "exclude": [
          "src/**/*.js",
        ],
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
          "src/**/*.ts",
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/compiled/babel-loader",
            "options": {
              "presets": [
                [
                  "<WORKSPACE>/packages/cli/babel-preset-app/dist/cjs/index.js",
                  {
                    "appDirectory": "<ROOT>",
                    "target": "client",
                    "useBuiltIns": "entry",
                    "useTsLoader": true,
                    "userBabelConfig": undefined,
                  },
                ],
              ],
            },
          },
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/ts-loader/index.js",
            "options": {
              "allowTsInNodeModules": true,
              "compilerOptions": {
                "module": "esnext",
                "target": "esnext",
              },
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/tsLoader > should set ts-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/compiled/babel-loader",
            "options": {
              "presets": [
                [
                  "<WORKSPACE>/packages/cli/babel-preset-app/dist/cjs/index.js",
                  {
                    "appDirectory": "<ROOT>",
                    "target": "client",
                    "useBuiltIns": "entry",
                    "useTsLoader": true,
                    "userBabelConfig": undefined,
                  },
                ],
              ],
            },
          },
          {
            "loader": "<WORKSPACE>/node_modules/<PNPM_INNER>/ts-loader/index.js",
            "options": {
              "allowTsInNodeModules": true,
              "compilerOptions": {
                "module": "esnext",
                "target": "esnext",
              },
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
}
`;
